### ISSUE TITLE
Can't specify `account_id` for some resources, but specifying it at the provider is deprecated.

### ISSUE THREAD

**Issue by MrJoy** [cloudflare/terraform-provider-cloudflare#1836]
### Confirmation

- [X] My issue isn't already found on the issue tracker.
- [X] I have replicated my issue using the latest version of the provider and it is still present.

### Terraform and Cloudflare provider version

```
Terraform v1.2.7
on darwin_arm64
+ provider registry.terraform.io/cloudflare/cloudflare v3.21.0
```

### Affected resource(s)

* `cloudflare_worker_script`

### Terraform configuration files

```hcl
N/A
```

### Debug output

N/A

### Panic output

N/A

### Expected output

No deprecation warnings.

### Actual output

When specifying the account ID via env var, or `account_id` attribute on the provider block, I get this:

```
│ Warning: Argument is deprecated
│
│   with provider["registry.terraform.io/cloudflare/cloudflare"],
│   on _provider.tf line 26, in provider "cloudflare":
│   26:   account_id = local.account-id
│
│ Use resource specific `account_id` attributes instead.
```

However, I have no apparent way to specify `account_id` for `cloudflare_worker_script` resources, as they have no such attribute.  Such attribute exists for some resources such as `cloudflare_zone`, and is inferred for associated resources like `cloudflare_zone_settings_override`.

Is there a way to Terraform workers that is clear of any deprecations, and I'm just missing it?

### Steps to reproduce

N/A

**Comment by jacobbednarz** [comment]
deprecating this is already tracked via #1646. you can follow that for progress on the deprecation.

**Comment by MrJoy** [comment]
I'm glad to hear that you're on top of the issue.  That said, the fact that CloudFlare shipped code that tells me my code is wrong, when there's no way to correct my code tends to shake my confidence in the CloudFlare Terraform provider.  It's not the first such occurrence to do so, either.

Frankly, I'm left feeling like this provider is just... not a priority for your organization, and that I'm taking a risk by using it.

**Comment by eherot** [comment]
@MrJoy This leaves me concerned as well. The point of a deprecation warning is to give users an opportunity to prepare for the future removal of functionality by shifting to a new functionality _before_ the current functionality is removed. The fact that folks complaining about this (useless) deprecation warning are just being sent to the 4.x milestone tracker makes me very concerned that CloudFlare is just going to forego the normal deprecation process altogether. I guess that's one way to do it, but it does beg the question of why we are being forced to suffer this warning that we cannot heed.

**Comment by jacobbednarz** [comment]
> That said, the fact that CloudFlare shipped code that tells me my code is wrong

Nothing tells you the code is wrong. The provider is issuing a warning that there is an upcoming deprecation and you should start making the swap before the functionality is removed entirely. It doesn't halt execution, it doesn't prevent you from doing anything. It simply says "the attribute you are using is going away, here is how to address it". If you bothered to read the code, you will see we are still supporting both forms of providing the account ID at the moment so if you want, you're free to ignore the guidance and just address it in the next major release. 

> ...and that I'm taking a risk by using it.

The project is open source. You are free to (at any moment in time) fork, modify and/or run your own version of it without any of your perceived risk. The risk you are talking about here is someone else maintaining the code for you 😉 if you don't feel comfortable with that, I'm not going to try and convince you otherwise. 

> The point of a deprecation warning is to give users an opportunity to prepare for the future removal of functionality by shifting to a new functionality before the current functionality is removed.

@eherot can you point out where are currently not doing this? As I mentioned above we still support both approaches and 4.x (the next major version which still has plenty of work to come) will consolidate to only supporting the new approach.  

> ...but it does beg the question of why we are being forced to suffer this warning that we cannot heed.

This is answered in the upstream PRs in cloudflare-go (that Terraform uses) but I will reiterate here. The problem is that `api.AccountID` was being used across client calls as a global configuration option and led to some very difficult to track issues and resource allocation. The only way we could address this in the SDK cleanly without having some resources broken for an unknown amount of time, was to rip it out there and shim in a fix for the provider. This isn't ideal however, it was the best of the options that end users had a deprecation notice before they could swap to it is pretty benign. The work to upgrade all resources isn't a simple task due to some internal changes so they are being tracked against the larger 4.x changes.

**Comment by eherot** [comment]
So it turned out that I was confused and that I was seeing this warning because I had `CLOUDFLARE_ACCOUNT_ID` set as an environment variable, which in turn sets `account_id` [at the provider level](https://github.com/cloudflare/terraform-provider-cloudflare/blob/master/internal/provider/provider.go#L128). It's my fault for forgetting that this variable was being set, however a mention of this default behavior in the deprecation notice would have saved me a lot of digging.

**Comment by jacobbednarz** [comment]
@eherot I'm unsure if there is a way to differentiate (even in the TF_LOG output) as the schema uses it for `DefaultFuncEnv`. This has come up before so I can have a look but can't promise anything.

**Comment by MrJoy** [comment]
> Nothing tells you the code is wrong. 

I apologize for my poor choice of words.  That said:  At the end of the day a deprecation warning is a call to action and I am unable to take action here.  Lecturing me about what deprecation means is condescending, unhelpful, and ignores the point I was trying to make.

> If you bothered to read the code, you will see we are still supporting both forms of providing the account ID at the moment so if you want, you're free to ignore the guidance and just address it in the next major release.

According to the ticket you yourself referenced, the provider does _not_ uniformly support both means of providing account ID yet.  Reading through [the code](https://github.com/cloudflare/terraform-provider-cloudflare/blob/d63c339c13b5a8c06bc09523a25c4ef71253f3f4/internal/provider/resource_cloudflare_worker_script.go), I see that there is no `account_id` field on the worker script resource type, as of d63c339.  And, of course, reading the code wasn't really needed here since I did in fact try to update to the new pattern and ran into the absence of the `account_id` field.

And that's my point: This transition could've been handled in a way that's more consistent with the CloudFlare brand image of expertise, conscientiousness, and thoroughness.  For example:  If the new mechanism had been introduced first -- gradually, if need be -- with the deprecation warning added only once the new mechanism is actually usable uniformly, that would have been much more consistent with what I've come to expect from CloudFlare.

Or even just including something in the message linking to #1646 so that when someone runs into the issue they're not left wondering what they're doing wrong that they're unable to actually resolve the deprecation warning and having to dig through the issues database in the hopes of finding the matter.

**Comment by eherot** [comment]
@jacobbednarz Yeah I figured the only way to do this would either be to make the deprecation notice very verbose and hint at it as a possibility (unless there is a way to use ternary syntax to construct the deprecation warning), write some kind of custom deprecation handler (yuck), or change the way the env var is treated and use it to set the default account ID on each resource individually (this would be ideal for my setup but possibly very different from what's intended for 4.x).

**Comment by myobie** [comment]
If you arrive here, like I did, and find deprecation warning unacceptable and will remove them one way or another... what I had to do is:

1. Leave `account_id` on the provider. Removing it straight away will blow things up.
2. Add an `account_id` to every resource that accepts it
3. Do a fully apply (apply -refresh-only does not cut it)
4. Then I was able to remove the `account_id` from the provider to finally stop having the deprecation warning
5. 💸 

Trying to remove the `account_id` from the provider before manually adding the `account_id` field to every resource and doing a full apply produced many super weird errors. 

Good luck, and I hope you don't waste as much time as I did trying to figure this out.
